src/kernel/include/inv.h

```diff
@@ -303,28 +303,36 @@ sinv_call(struct thread *thd, struct cap_sinv *sinvc, struct pt_regs *regs, stru
 	return;
 }
 
-static inline void
+static inline int
 sret_ret(struct thread *thd, struct pt_regs *regs, struct cos_cpu_local_info *cos_info)
 {
-	struct comp_info *ci;
-	unsigned long     ip, sp;
+	struct comp_info *       ci;
+	unsigned long            ip = 0;
+	unsigned long            sp = 0;
+	unsigned long            in_fault = 0;
 
-	ci = thd_invstk_pop(thd, &ip, &sp, cos_info);
+	ci = thd_invstk_pop(thd, &ip, &sp, &in_fault, cos_info);
 	if (unlikely(!ci)) {
 		__userregs_set(regs, 0xDEADDEAD, 0, 0);
-		return;
+		goto ret;
 	}
 
 	if (unlikely(!ltbl_isalive(&ci->liveness))) {
 		printk("cos: ret comp (liveness %d) doesn't exist!\n", ci->liveness.id);
 		// FIXME: add fault handling here.
 		__userregs_set(regs, -EFAULT, __userregs_getsp(regs), __userregs_getip(regs));
-		return;
+		goto ret;
 	}
 
 	pgtbl_update(ci->pgtbl);
 	/* Set return sp and ip and function return value in eax */
 	__userregs_set(regs, __userregs_getinvret(regs), sp, ip);
+ret:
```

Comment:
Wait.. I thought you said we're saving a branch in the fast-path..? But we're not here.